File: Sources/SkipLibKt/Skip/Globals.kt
// Copyright 2023 Skip
//
// This is free software: you can redistribute and/or modify it
// under the terms of the GNU Lesser General Public License 3.0
// as published by the Free Software Foundation https://fsf.org
package skip.lib

fun fatalError(message: String = "fatalError"): Nothing = error(message)
fun assertionFailure(message: String = "assertionFailure"): Nothing = error(message)
fun preconditionFailure(message: String = "preconditionFailure"): Nothing = error(message)
fun precondition(condition: Boolean, message: String = "precondition"): Unit = require(condition, { message })

typealias Never = Nothing

fun type(of: Any): kotlin.reflect.KClass<*> = of::class

fun <T> swap(a: InOut<T>, b: InOut<T>) {
    val t = a.value
    a.value = b.value
    b.value = t
}

fun <T : Comparable<T>> min(a: T, b: T): T {
	return if (a <= b) a else b
}

fun <T : Comparable<T>> max(a: T, b: T): T {
	return if (a >= b) a else b
}

fun round(x: Double): Double {
	return kotlin.math.round(x)
}

fun round(x: Float): Float {
	return kotlin.math.round(x)
}

class NullReturnException: Exception() {
}

fun print(vararg args: Any?, separator: String = " ", terminator: String = "\n") {
	args.forEachIndexed { index, arg ->
		if (index > 0) kotlin.io.print(separator)
		kotlin.io.print(arg)
	}
	kotlin.io.print(terminator)
}

fun debugPrint(vararg args: Any?, separator: String = " ", terminator: String = "\n") = print(args, separator, terminator)

File: Sources/SkipLib/Globals.swift
// Copyright 2023 Skip
//
// This is free software: you can redistribute and/or modify it
// under the terms of the GNU Lesser General Public License 3.0
// as published by the Free Software Foundation https://fsf.org

// THIS TRANSPILATION IS NOT USED
//
// This file only exists to provide symbols for implemented API to the transpiler.
//

#if SKIP

/// Unconditionally prints a given message and stops execution.
public func fatalError(message: String = "fatalError") -> Never {
    Swift.fatalError()
}

/// Performs a traditional C-style assert with an optional message.
public func assert(_ condition: Bool, _ message: String? = nil) {
    fatalError()
}

/// Checks a necessary condition for making forward progress.
public func precondition(_ condition: Bool, _ message: String? = nil) {
    fatalError()
}

/// Indicates that an internal sanity check failed.
@inlinable public func assertionFailure(_ message: String? = nil) {
    fatalError()
}

/// Indicates that a precondition was violated.
public func preconditionFailure(_ message: String? = nil) -> Never {
    fatalError()
}

public func type(of: Any) -> Any.Type {
    fatalError()
}

public func swap<T>(_ a: inout T, _ b: inout T) {
    fatalError()
}

public func min<T>(_ x: T, _ y: T) -> T where T : Comparable {
    fatalError()
}

public func max<T>(_ x: T, _ y: T) -> T where T : Comparable {
    fatalError()
}

public func round(_ x: Double) -> Double {
    fatalError()
}

public func round(_ x: Float) -> Float {
    fatalError()
}

#endif
